lib/ansible/modules/remote_management/ucs/ucs_dns_server.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: ucs_dns_server
+short_description: Configures DNS server on Cisco UCS Manager
+extends_documentation_fragment:
+- ucs
+description:
+- Configures DNS server on Cisco UCS Manager.
+- Examples can be used with the L(UCS Platform Emulator,https://communities.cisco.com/ucspe).
+options:
+  state:
+    description:
+    - If C(absent), will remove a DNS server.
+    - If C(present), will add or update a DNS server.
+    choices: [absent, present]
+    default: present
+
+  dns_server:
+    description:
+    - DNS server IP address.
+    - Enter a valid IPV4 Address.
+    aliases: [ name ]
+    default: ""
+
+  description:
+    description:
+    - A user-defined description of the DNS server.
+    - Enter up to 256 characters.
+    - "You can use any characters or spaces except the following:"
+    - "` (accent mark), \ (backslash), ^ (carat), \" (double quote), = (equal sign), > (greater than), < (less than), or ' (single quote)."
+    aliases: [ descr ]
+    default: ""
+
+requirements:
+- ucsmsdk
+author:
+- John McDonough (@movinalot)
+- CiscoUcs (@CiscoUcs)
+version_added: "2.7"
+'''
+
+EXAMPLES = r'''
+- name: Configure DNS server
+  ucs_dns_server:
+    hostname: 172.16.143.150
+    username: admin
+    password: password
+    dns_server: 10.10.10.10
+    description: DNS Server IP address
+    state: present
+
+- name: Remove DNS server
+  ucs_dns_server:
+    hostname: 172.16.143.150
+    username: admin
+    password: password
+    dns_server: 10.10.10.10
+    state: absent
+'''
+
+RETURN = r'''
+#
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.remote_management.ucs import UCSModule, ucs_argument_spec
+
+
+def run_module():
+    argument_spec = ucs_argument_spec
+    argument_spec.update(
+        dns_server=dict(type='str', aliases=['name']),
+        description=dict(type='str', aliases=['descr'], default=''),
+        state=dict(type='str', default='present', choices=['present', 'absent']),
+    )
+
+    module = AnsibleModule(
+        argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ['state', 'present', ['dns_server']],
+        ],
+    )
+    # UCSModule verifies ucsmsdk is present and exits on failure.  Imports are below ucs object creation.
+    ucs = UCSModule(module)
+
+    err = False
+
+    from ucsmsdk.mometa.comm.CommDnsProvider import CommDnsProvider
+
+    changed = False
+    try:
```

Comment:
@dsoper2 I think I mentioned it once before, but this one big try-except block you are using in almost every module is considered a bad practice. We tend to prefer small try-except blocks and catching specific exceptions with specific error-messages guiding users.

I understand that that's not always the easiest path, because you may not always know in advance in what ways something can break, but that's something you learn along the way.

In this example, there are a lot if statements that can't go wrong (like assignments and conditionals). On the other hand, you're calling class-methods that might as well catch their own exceptions, so there's no need to catch them in the module.

Maybe the only call you want to add in a try-except block is the call to CommDnsProvider, because it's external you may need the handle raised exceptions.

Reply:
I would ask that you merge this as is and as a group, we come up with a framework that can be applied to all modules in future revisions.  Since this module is so similar to ucs_ntp_server I would like to keep their structures very similar.